lib/debugger/debugger.dart

```diff
@@ -0,0 +1,1309 @@
+    _executionPointElement = null;
+
+    if (script.source != null && position != null) {
+      _executionPointElement =
+          span(c: 'octicon octicon-arrow-up execution-marker');
+
+      codeMirror.addWidget(
+        new Position(position.line - 1, position.column - 1),
+        _executionPointElement.element,
+      );
+    }
+  }
+
+  void clearExecutionPoint() {
+    executionPoint = null;
+    _clearLineClass();
+    _refreshMarkers();
+  }
+
+  final Map<String, int> _lastScrollPositions = <String, int>{};
+
+  void displayScript(Script newScript, {Pos scrollTo}) {
+    if (currentScript != null) {
+      final ScrollInfo scrollInfo = codeMirror.getScrollInfo();
+      _lastScrollPositions[currentScript.uri] = scrollInfo.top;
+    }
+
+    final bool sameScript = currentScript?.uri == newScript?.uri;
+
+    currentScript = newScript;
+
+    if (newScript == null) {
+      codeMirror.getDoc().setValue('');
+    } else {
+      // TODO(devoncarew): set the mode to either dart or javascript
+      // codeMirror.setMode(mode);
+
+      if (!sameScript) {
+        final String source = newScript?.source ?? '<source not available>';
+        codeMirror.getDoc().setValue(source);
+      }
+
+      if (scrollTo != null) {
+        codeMirror.scrollIntoView(scrollTo.line - 1, 0, margin: 150);
+      } else {
+        final int top = _lastScrollPositions[newScript.uri] ?? 0;
+        codeMirror.scrollTo(0, top);
+      }
+    }
+
+    _executionPointElement?.dispose();
+    _executionPointElement = null;
+
+    _refreshMarkers();
+  }
+}
+
+typedef URIDescriber = String Function(String uri);
+
+class BreakpointsView {
+  BreakpointsView(this._breakpointsCountDiv, DebuggerState debuggerState,
+      URIDescriber uriDescriber) {
+    _items = new SelectableList<Breakpoint>()
+      ..flex()
+      ..clazz('menu-item-bottom-border')
+      ..element.style.overflowY = 'scroll';
+
+    _items.setRenderer((Breakpoint breakpoint) {
+      final dynamic location = breakpoint.location;
+
+      final CoreElement element = li(text: '', c: 'list-item');
+
+      if (location is UnresolvedSourceLocation) {
+        element.text = uriDescriber(location.script.uri);
+        element.add(span(text: ' line ${location.line}', c: 'subtle'));
+      } else if (location is SourceLocation) {
+        element.text = uriDescriber(location.script.uri);
+
+        // Modify the rendering slightly asynchronously.
+        debuggerState.getScript(location.script).then((Script script) {
+          final Pos pos =
+              debuggerState.calculatePosition(script, location.tokenPos);
+          element.add(span(text: ' line ${pos.line}', c: 'subtle'));
+        });
+      }
+
+      if (!breakpoint.resolved) {
+        element.add(span(text: ' (unresolved)', c: 'subtle'));
+      }
+
+      return element;
+    });
+  }
+
+  Stream<Breakpoint> get onDoubleClick => _items.onDoubleClick;
+
+  final CoreElement _breakpointsCountDiv;
```

Comment:
consider moving all fields up closer to the top of the class.

Reply:
done